### PR TITLE
EN-65713: separate docker build from publish

### PR DESCRIPTION
**Pull request by sarah-salinas**
This commit moves the release tagging after a successful docker build.  In order to do this, it separates the docker build from the publishing of the images into separate Jenkinsfile steps.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -21,6 +21,7 @@ def releaseTag = new com.socrata.ReleaseTag(steps, service_server)
 pipeline {
   options {
     ansiColor('xterm')
+    disableConcurrentBuilds(abortPrevious: true)
```
Comment (sarah-salinas) [inline Jenkinsfile:24]: Bonus: helpful setting that we now use on other services.  If a new build starts while an existing build is in progress, the first build is cancelled.

File: Jenkinsfile [hunk 2]
```diff
@@ -70,103 +55,146 @@ pipeline {
         }
       }
     }
-    stage('Dockerize SoQL Server PG') {
+    stage('Docker Build') {
```
Comment (sarah-salinas) [inline Jenkinsfile:58]: I added top level stages for "Docker Build" and "Publish".  This allows us to share the `when` condition and run them in parallel.

File: Jenkinsfile [hunk 3]
```diff
@@ -70,103 +55,146 @@ pipeline {
...
         }
       }
     }
-    stage('Dockerize Secondary') {
+    stage('Publish') {
       when {
-        not { expression { isPr } }
+        allOf {
+          not { expression { isPr } }
+          not { expression { return params.RELEASE_BUILD && params.RELEASE_DRY_RUN } }
```
Comment (sarah-salinas) [inline Jenkinsfile:117]: We exclude release dry runs from the publish stage.

File: Jenkinsfile [hunk 4]
```diff
@@ -70,103 +55,146 @@ pipeline {
...
-            env.SECONDARY_DOCKER_TAG = dockerize_secondary.docker_build_specify_tag_and_push(params.RELEASE_NAME, sbtbuild.getDockerPath(project_wd_secondary), env.STORE_PG_ARTIFACT, env.REGISTRY_PUSH)
-          } else {
-            env.REGISTRY_PUSH = 'internal'
-            env.SECONDARY_DOCKER_TAG = dockerize_secondary.docker_build('STAGING', env.GIT_COMMIT, sbtbuild.getDockerPath(project_wd_secondary), env.STORE_PG_ARTIFACT, env.REGISTRY_PUSH)
+      parallel {
+        stage('Publish - SoQL Server') {
+          steps {
+            script {
+              lastStage = env.STAGE_NAME
+              Set environments = (params.RELEASE_BUILD) ? [] : ['internal']
```
Comment (sarah-salinas) [inline Jenkinsfile:125]: Passing `[]` to overrideEnvironments will not override the environments, so the docker image will be published to all registries.

File: Jenkinsfile [hunk 5]
```diff
@@ -70,103 +55,148 @@ pipeline {
...
-          } else {
-            env.REGISTRY_PUSH = 'internal'
-            env.SECONDARY_DOCKER_TAG = dockerize_secondary.docker_build('STAGING', env.GIT_COMMIT, sbtbuild.getDockerPath(project_wd_secondary), env.STORE_PG_ARTIFACT, env.REGISTRY_PUSH)
+      parallel {
+        stage('Publish - SoQL Server') {
+          steps {
+            script {
+              lastStage = env.STAGE_NAME
+              // Passing [] to overrideEnvironments will not override the environments, so the docker image will be published to all environments
+              Set environments = (params.RELEASE_BUILD) ? [] : ['internal']
```
Comment (sarah-salinas) [inline Jenkinsfile:126]: This solution makes the code DRY, and doesn't duplicate the full list of environments here.

File: Jenkinsfile [hunk 6]
```diff
@@ -70,103 +55,148 @@ pipeline {
...
       }
       stages {
-        stage('Deploy SoQL Server PG') {
+        stage('Deploy - SoQL Server') {
           steps {
             script {
               lastStage = env.STAGE_NAME
-              // uses env.DOCKER_TAG and deploys to staging by default
-              marathonDeploy(serviceName: env.DEPLOY_PATTERN, waitTime: '60')
+              marathonDeploy(serviceName: env.DEPLOY_PATTERN, tag: env.DOCKER_TAG, environment: 'staging', waitTime: '60')
```
Comment (sarah-salinas) [inline Jenkinsfile:191]: It's a better practice to be explicit about parameters.

**Comment by sarah-salinas** [comment]
## Testing

### Simulated Main Build

https://jenkins-build.socrata.com/job/soql-postgres-adapter-branch/103/console

### Release Dry Run



### Simulated Release Build

**Review by rjmac — APPROVED** [review]

File: Jenkinsfile [hunk 7]
```diff
@@ -21,13 +21,14 @@ def releaseTag = new com.socrata.ReleaseTag(steps, service_server)
 pipeline {
   options {
     ansiColor('xterm')
+    disableConcurrentBuilds(abortPrevious: true)
   }
   parameters {
+    string(name: 'AGENT', defaultValue: 'build-worker', description: 'Which build agent to use?')
```
Comment (sarah-salinas) [inline Jenkinsfile:27]: We typically list the agent and branch specifier first.

File: Jenkinsfile [hunk 8]
```diff
@@ -70,103 +55,189 @@ pipeline {
...
-          } else {
-            env.REGISTRY_PUSH = 'internal'
-            env.DOCKER_TAG = dockerize_server.docker_build('STAGING', env.GIT_COMMIT, sbtbuild.getDockerPath(project_wd_server), sbtbuild.getDockerArtifact(project_wd_server), env.REGISTRY_PUSH)
+      parallel {
+        stage ('Docker Build - SoQL Server') {
+          steps {
+            script {
+              lastStage = env.STAGE_NAME
+              if (params.RELEASE_BUILD) {
+                env.DOCKER_TAG = dockerize_server.dockerBuildWithSpecificTag(
```
Comment (sarah-salinas) [inline Jenkinsfile:68]: New methods dockerBuildWithSpecificTag and dockerBuildWithDefaultTag (see https://github.com/socrata/pipelines/blob/main/src/com/socrata/Dockerize.groovy)
For each one, the parameter is a map, which can be entered without the outside brackets.
They return just the docker tag (`H24` or `staging_183_289dkkck`).

File: Jenkinsfile [hunk 9]
```diff
@@ -70,103 +55,189 @@ pipeline {
...
       stages {
-        stage('Deploy SoQL Server PG') {
+        stage('Deploy - SoQL Server') {
           steps {
             script {
               lastStage = env.STAGE_NAME
-              // uses env.DOCKER_TAG and deploys to staging by default
-              marathonDeploy(serviceName: env.DEPLOY_PATTERN, waitTime: '60')
+              marathonDeploy(
+                serviceName: env.DEPLOY_PATTERN,
```
Comment (sarah-salinas) [inline Jenkinsfile:224]: It's a better practice to be explicit about parameters and put them on their own line for readability and git history clarity.

**Comment by sarah-salinas** [comment]
## Retesting

### Simulated main build

https://jenkins-build.socrata.com/job/soql-postgres-adapter-branch/107/console

### Release Dry Run


### Simulated Release Build

File: Jenkinsfile (hunk 8 above)
Comment (sarah-salinas) [reply Jenkinsfile:70]: See https://github.com/socrata/activity-feed/pull/188 for example usage.

File: Jenkinsfile [hunk 10]
```diff
@@ -70,103 +55,189 @@ pipeline {
...
-          } else {
-            env.REGISTRY_PUSH = 'internal'
-            env.SECONDARY_DOCKER_TAG = dockerize_secondary.docker_build('STAGING', env.GIT_COMMIT, sbtbuild.getDockerPath(project_wd_secondary), env.STORE_PG_ARTIFACT, env.REGISTRY_PUSH)
+      parallel {
+        stage('Publish - SoQL Server') {
+          steps {
+            script {
+              lastStage = env.STAGE_NAME
+              if (params.RELEASE_BUILD) {
+                env.BUILD_ID = dockerize_server.publish(sourceTag: env.DOCKER_TAG)
```
[the remark this answers is not in the document]
Comment (sarah-salinas) [reply Jenkinsfile:143]: See https://github.com/socrata/activity-feed/pull/188 for example usage.

File: Jenkinsfile [hunk 11]
```diff
@@ -21,13 +21,16 @@ def releaseTag = new com.socrata.ReleaseTag(steps, service_server)
 pipeline {
   options {
     ansiColor('xterm')
+    buildDiscarder(logRotator(numToKeepStr: '50'))
+    disableConcurrentBuilds(abortPrevious: true)
+    timeout(time: 70, unit: 'MINUTES')
```
Comment (sarah-salinas) [inline Jenkinsfile:26]: Deploy of soql-server is set to max out at 60 min.  The rest of the build usually takes 6 min at most.

File: Jenkinsfile [hunk 12]
```diff
@@ -21,13 +21,16 @@ def releaseTag = new com.socrata.ReleaseTag(steps, service_server)
 pipeline {
   options {
     ansiColor('xterm')
+    buildDiscarder(logRotator(numToKeepStr: '50'))
+    disableConcurrentBuilds(abortPrevious: true)
+    timeout(time: 100, unit: 'MINUTES')
```
Comment (sarah-salinas) [inline Jenkinsfile:26]: This gives a reasonable time for killing a stuck build.  Deploy of soql-server is set to max out at 60 min. The rest of the build usually takes 6 min at most.

File: Jenkinsfile [hunk 13]
```diff
@@ -21,13 +21,16 @@ def releaseTag = new com.socrata.ReleaseTag(steps, service_server)
 pipeline {
   options {
     ansiColor('xterm')
+    buildDiscarder(logRotator(numToKeepStr: '50'))
```
Comment (sarah-salinas) [inline Jenkinsfile:24]: We typically use 20 or 50.

**Review by rjmac — APPROVED** [review]

File: Jenkinsfile (hunk 1 above)
Comment (ds2643) [reply Jenkinsfile:25]: Is the premature termination always safe?

File: Jenkinsfile (hunk 7 above)
Comment (ds2643) [reply Jenkinsfile:29]: Very minor: The description doesn't seem like a question.

File: Jenkinsfile [hunk 14]
```diff
@@ -70,103 +57,190 @@ pipeline {
...
-                rmsSupportedEnvironment.production
-              )
-            }
+            currentBuild.description = "${env.BUILD_ID} & ${env.SECONDARY_BUILD_ID}"
           }
         }
       }
     }
-    stage('Deploys:') {
+    stage('Deploy:') {
```
Comment (ds2643) [inline Jenkinsfile:214]: Should the colon be removed to match the formatting of other stage names?

File: Jenkinsfile [hunk 15]
```diff
@@ -70,103 +57,190 @@ pipeline {
...
+            success {
+              script {
+                if (params.RELEASE_BUILD) {
+                  Map buildInfo = [
+                    "project_id": service_secondary,
+                    "build_id": env.SECONDARY_BUILD_ID,
+                    "release_id": params.RELEASE_NAME,
+                    "git_tag": env.GIT_TAG
+                  ]
+                  createBuild(
```
Comment (ds2643) [inline Jenkinsfile:196]: With respect to notifying RMS, what happens if one of the artifacts (soql-server, secondary) fails and the other succeeds?

**Review by ds2643 — APPROVED** [review]
These changes largely reflect similar changes across other repositories. Overall, the refactoring plays an important role in simplifying the question of what defines success in the context of RMS tracked builds.

File: Jenkinsfile (hunk 15 above)
Comment (sarah-salinas) [reply Jenkinsfile:196]: The stages are serial, so if there is a failure in the docker build stage, then it won't notify RMS.  
If publishing soql-server succeeds, but not secondary, then we have code problems in the pipelines repo.  Tactically, we can rebuild with `replay` and select which stages to run.

File: Jenkinsfile (hunk 1 above)
Comment (sarah-salinas) [reply Jenkinsfile:25]: Builds have failed in many places and it's never been unsafe or required additional cleanup.

**Pull request merged**